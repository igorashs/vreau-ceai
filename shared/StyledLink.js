import styled from 'styled-components';
import { Button } from './Button';
import Link from 'next/link';

const StyledA = styled.a`
  text-decoration: none;
  display: inline-flex;
  align-items: center;
  gap: 7px;

  :hover {
    filter: brightness(0.7);
  }

  ${({ accent }) => {
    switch (accent) {
      case 'light':
        return 'color: var(--accent-text-light);';
      case 'dark':
        return 'color: var(--accent-text-dark);';
      default:
        return 'color: var(--text-light);';
    }
  }}

  ${({ underline }) => `text-decoration: ${underline ? 'underline' : 'none'};`}
`;

export const StyledLink = ({
  href,
  label,
  text,
  accent,
  Icon,
  target,
  rel,
  underline,
  button = false,
  children
}) => (
  <Link href={href} passHref>
    {button ? (
      <Button as="a" icon={!text} aria-label={label} target={target} rel={rel}>
        {Icon && <Icon />}
        {text}
      </Button>
    ) : (
      <StyledA
        aria-label={label}
        accent={accent}
        target={target}
        rel={rel}
        underline={underline}
      >
        {Icon && <Icon />}
        {text || children}
      </StyledA>
    )}
  </Link>
);
