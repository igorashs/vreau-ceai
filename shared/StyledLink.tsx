import styled from 'styled-components';
import Link from 'next/link';
import Button from './Button';

interface StyledAProps {
  accent?: 'light' | 'dark';
  underline?: boolean;
}

const StyledA = styled.a<StyledAProps>`
  text-decoration: none;
  display: inline-flex;
  align-items: center;
  gap: 7px;

  :hover {
    filter: brightness(0.7);
  }

  ${({ accent }) => {
    switch (accent) {
      case 'light':
        return 'color: var(--accent-text-light);';
      case 'dark':
        return 'color: var(--accent-text-dark);';
      default:
        return 'color: var(--text-light);';
    }
  }}

  ${({ underline }) => `text-decoration: ${underline ? 'underline' : 'none'};`}
`;

interface StyledLinkProps {
  href: string;
  label: string;
  text?: string;
  accent?: 'light' | 'dark';
  Icon?: React.FC<React.SVGProps<SVGElement>>;
  target?: string;
  rel?: string;
  underline?: boolean;
  button?: boolean;
}

export const StyledLink = ({
  href,
  label,
  text = '',
  accent,
  Icon,
  target = '',
  rel = '',
  underline = false,
  button = false,
  children,
}: React.PropsWithChildren<StyledLinkProps>) => (
  <Link href={href} passHref>
    {button ? (
      <Button as="a" icon={!text} aria-label={label} target={target} rel={rel}>
        {Icon && <Icon />}
        {text}
      </Button>
    ) : (
      <StyledA
        aria-label={label}
        accent={accent}
        target={target}
        rel={rel}
        underline={underline}
      >
        {Icon && <Icon />}
        {text || children}
      </StyledA>
    )}
  </Link>
);
