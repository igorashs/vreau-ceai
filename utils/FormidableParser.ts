import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
