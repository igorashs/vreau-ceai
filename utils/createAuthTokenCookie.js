import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
