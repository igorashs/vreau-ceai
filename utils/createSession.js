import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
