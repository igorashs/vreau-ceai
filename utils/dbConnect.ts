import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
