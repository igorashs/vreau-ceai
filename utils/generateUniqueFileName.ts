import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
