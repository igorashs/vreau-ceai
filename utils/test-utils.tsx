import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
