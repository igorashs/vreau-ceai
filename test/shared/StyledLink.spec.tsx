import StyledLink from '@/shared/StyledLink';
import { render } from '@/utils/test-utils';

describe('StyledLink', () => {
  it('renders correctly', () => {
    const { container, getByRole, getByText } = render(
      <StyledLink href="/home" label="go home">
        home
      </StyledLink>,
    );

    expect(getByRole('link')).toBeInTheDocument();
    expect(getByText('home')).toBeVisible();
    expect(container.firstChild).toHaveAttribute('aria-label', 'go home');
    expect(container.firstChild).toHaveAttribute('href', '/home');
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('button prop', () => {
    it('renders with button styles', () => {
      const { getByRole, getByText, queryByText } = render(
        <StyledLink href="/home" label="go home" button text="home">
          children
        </StyledLink>,
      );

      expect(getByRole('link')).toBeInTheDocument();
      expect(getByText('home')).toBeVisible();
      expect(queryByText('children')).not.toBeInTheDocument();
    });
  });

  describe('accent prop', () => {
    it('renders with light accent', () => {
      const { container } = render(
        <StyledLink href="/home" label="go home" accent="light">
          home
        </StyledLink>,
      );

      expect(container.firstChild).toHaveStyleRule(
        'color',
        'var(--accent-text-light)',
      );
    });

    it('renders with dark accent', () => {
      const { container } = render(
        <StyledLink href="/home" label="go home" accent="dark">
          home
        </StyledLink>,
      );

      expect(container.firstChild).toHaveStyleRule(
        'color',
        'var(--accent-text-dark)',
      );
    });
  });

  describe('underline prop', () => {
    it('renders with underline rule', () => {
      const { container } = render(
        <StyledLink href="/home" label="go home" underline>
          home
        </StyledLink>,
      );

      expect(container.firstChild).toHaveStyleRule(
        'text-decoration',
        'underline',
      );
    });
  });
});
