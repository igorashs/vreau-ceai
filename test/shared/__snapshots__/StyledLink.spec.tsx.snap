// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyledLink renders correctly 1`] = `
.c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 7px;
  color: var(--text-light);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:hover {
  -webkit-filter: brightness(0.7);
  filter: brightness(0.7);
}

<a
  aria-label="go home"
  class="c0"
  href="/home"
  rel=""
  target=""
>
  home
</a>
`;
